python/ml4ir/base/model/scoring/monte_carlo_scorer.py

```diff
@@ -99,12 +91,171 @@ def __init__(
                          logs_dir=logs_dir,
                          **kwargs)
 
-        self.monte_carlo_test_trials = self.model_config[MonteCarloInferenceKey.MONTE_CARLO_TRIALS][MonteCarloInferenceKey.NUM_TEST_TRIALS]
-        self.monte_carlo_training_trials = self.model_config[MonteCarloInferenceKey.MONTE_CARLO_TRIALS][MonteCarloInferenceKey.NUM_TRAINING_TRIALS]
+        self.use_fixed_mask_in_training = bool(self.model_config[MonteCarloInferenceKey.MONTE_CARLO_TRIALS].get(
+            MonteCarloInferenceKey.USE_FIXED_MASK_IN_TRAINING, False))
+        if not self.use_fixed_mask_in_training:
+            self.monte_carlo_training_trials = self.model_config[MonteCarloInferenceKey.MONTE_CARLO_TRIALS][
+                MonteCarloInferenceKey.NUM_TRAINING_TRIALS]
+            # Adding 1 here to account of the extra inference run with training=False.
+            self.monte_carlo_training_trials_tf = tf.constant(self.monte_carlo_training_trials + 1, dtype=tf.float32)
+
+        self.use_fixed_mask_in_testing = bool(self.model_config[MonteCarloInferenceKey.MONTE_CARLO_TRIALS].get(
+            MonteCarloInferenceKey.USE_FIXED_MASK_IN_TESTING, False))
+        if not self.use_fixed_mask_in_testing:
+            self.monte_carlo_test_trials = self.model_config[MonteCarloInferenceKey.MONTE_CARLO_TRIALS][
+                MonteCarloInferenceKey.NUM_TEST_TRIALS]
+            # Adding 1 here to account of the extra inference run with training=False.
+            self.monte_carlo_test_trials_tf = tf.constant(self.monte_carlo_test_trials + 1, dtype=tf.float32)
+
+        self.features_with_fixed_masks = self.model_config[MonteCarloInferenceKey.MONTE_CARLO_TRIALS].get(MonteCarloInferenceKey.FEATURES_WITH_FIXED_MASKS, [])
+        if self.features_with_fixed_masks:
+            self.features_with_fixed_masks = [f.strip() for f in self.features_with_fixed_masks.split()]
+            self.fixed_feature_count = len(self.features_with_fixed_masks)
+            self.fixed_mask = list(itertools.product([0, 1], repeat=self.fixed_feature_count))
+            self.monte_carlo_fixed_trials_count = tf.constant(len(self.fixed_mask), dtype=tf.float32)
```

Comment:
@mohazahran this feels hacky to me because so far MCScorer did not have any notion of features and it was being handled by the masking layers.

Reply:
@lastmansleeping I see your point, however for fixed masks feature masking are dependent. For example, we want to have all combination of the set of features, run inference and aggregate. 
If we want the masking layer do the masking, then it would need to create duplicates of each query as it cannot just simply apply one mask at a time as in the stochastic MC. Creating duplicates can mess up the training/eval pipeline. This made me move this in the scorer itself.